packages/ckeditor5-ui/src/editorui/accessibilityhelp.ts

```diff
@@ -0,0 +1,382 @@
+	 * 			keystroke: 'Tab'
+	 * 		}
+	 * 	]
+	 * } );
+	 * ```
+	 */
+	public registerKeystrokes( options: {
+		category?: string;
+		group?: string;
+		groupLabel?: string;
+		keystrokes: Array<AccessibilityHelpKeystrokeDefinition>;
+	} | Array<AccessibilityHelpKeystrokeDefinition> | AccessibilityHelpKeystrokeDefinition ): void {
+		if ( options instanceof Array ) {
+			options = {
+				keystrokes: options
+			};
+		} else if ( typeof options === 'object' && 'keystroke' in options ) {
+			options = {
+				keystrokes: [ options ]
+			};
+		}
+
+		const categoryId = options.category || DEFAULT_CATEGORY_ID;
+
+		if ( !this._keystrokes.has( categoryId ) ) {
+			/**
+			 * Cannot register keystrokes in an unknown category. Use
+			 * {@link module:ui/editorui/accessibilityhelp~AccessibilityHelp#registerKeystrokeCategory}
+			 * to register a new category or make sure the specified category exists.
+			 *
+			 * @error accessibility-help-unknown-category
+			 * @param categoryId The id of the unknown keystroke category.
+			 */
+			throw new CKEditorError( 'accessibility-help-unknown-category', { categoryId } );
+		}
+
+		const category = this._keystrokes.get( categoryId )!;
+		const groupId = options.group || DEFAULT_GROUP_ID;
+
+		if ( !category.groups.has( groupId ) && !options.groupLabel ) {
+			/**
+			 * Cannot register keystrokes in an unknown group.
+			 *
+			 * * If you want to create a new group, make sure you pass the `groupLabel` option to the
+			 * {@link module:ui/editorui/accessibilityhelp~AccessibilityHelp#registerKeystrokes} method.
+			 * * If you want to add keystrokes to an existing group, make sure it exists.
+			 *
+			 * @error accessibility-help-unknown-group
+			 * @param groupId The id of the unknown keystroke group.
+			 * @param categoryId The id of category the unknown group should belong to.
+			 */
+			throw new CKEditorError( 'accessibility-help-unknown-group', { groupId, categoryId } );
+		} else if ( options.groupLabel ) {
+			category.groups.set( groupId, {
+				label: options.groupLabel,
+				keystrokes: options.keystrokes
+			} );
+		} else {
+			category.groups.get( groupId )!.keystrokes.push( ...options.keystrokes );
+		}
+	}
+
+	/**
+	 * Injects a help text into each editing root's `aria-label` attribute allowing assistive technology users
+	 * to discover the availability of the Accessibility help dialog.
+	 */
+	private _setupRootLabels() {
+		const editor = this.editor;
+		const t = editor.t;
+
+		editor.on( 'ready', () => {
+			editor.editing.view.change( writer => {
+				for ( const rootName of editor.model.document.getRootNames() ) {
+					const viewRoot = editor.editing.view.document.getRoot( rootName );
+					const currentAriaLabel = viewRoot!.getAttribute( 'aria-label' );
+
+					writer.setAttribute(
+						'aria-label',
+						`${ currentAriaLabel }. ${ t( 'Press %0 for help.', [ getEnvKeystrokeText( 'Alt+0' ) ] ) }`,
+						viewRoot!
+					);
+				}
+			} );
+		} );
+	}
```

Comment:
I wish there was an API for that. It's not pretty, and besides, if there were more places like this one, maintaining order would become an issue.

Reply:
I'll take a look at it but I want to submit the review, so no comment here now.

Reply:
I addressed this in 6c4df0b1352bb02da3a95344020ec7daab5df77e.

It turned out the `add` event on `editor.editing.view.document.roots` fires too early to inject this kind of label (`InlineEditableUIView` hasn't been rendered yet).

For clarity, I don't like this piece of code either. AFAIR I proposed a better solution a couple of years ago but it was rejected. There's a lot of magic going on around this process:

1. first, `InlineEditableUIView` is created along the engine root editable element, and you can write its attributes using downcast writer (if you hook into the process too early),
2. But whatever you do, it will get erased as soon as the related `InlineEditableUIView` renders and [sets the attribute via UI library API](https://github.com/ckeditor/ckeditor5/blob/6c4df0b1352bb02da3a95344020ec7daab5df77e/packages/ckeditor5-ui/src/editableui/inline/inlineeditableuiview.ts#L66-L70)
3. Only after `InlineEditableUIView#render()`, you can extend its attributes for real using downast writer.

We really need to streamline this at some point.